subsys/net/lib/mqtt/mqtt_transport_websocket.c

```diff
@@ -0,0 +1,159 @@
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ */
+int mqtt_client_websocket_connect(struct mqtt_client *client)
+{
+	int transport_sock;
+	int ret;
+
+	if (client->transport.type == MQTT_TRANSPORT_NON_SECURE_WEBSOCKET) {
+		ret = mqtt_client_tcp_connect(client);
+		if (ret < 0) {
+			return ret;
+		}
+
+		transport_sock = client->transport.tcp.sock;
+	}
+#if defined(CONFIG_MQTT_LIB_TLS)
+	else if (client->transport.type == MQTT_TRANSPORT_SECURE_WEBSOCKET) {
+		ret = mqtt_client_tls_connect(client);
+		if (ret < 0) {
+			return ret;
+		}
+
+		transport_sock = client->transport.tls.sock;
+	}
+#endif
+	else {
+		return -EINVAL;
+	}
+
+	if (client->transport.websocket.config.service_url == NULL) {
+		client->transport.websocket.config.service_url = "/";
+	}
+
+	if (client->transport.websocket.config.server_name == NULL) {
+		client->transport.websocket.config.server_name = "localhost";
+	}
+
+	client->transport.websocket.sock = websocket_connect(
+		transport_sock,
+		client->transport.websocket.config.server_name,
+		client->transport.websocket.config.service_url,
+		NULL, NULL, NULL,
+		client->transport.websocket.config.temp_recv_buf,
+		sizeof(client->transport.websocket.config.temp_recv_buf),
+		K_FOREVER, NULL);
+	if (client->transport.websocket.sock < 0) {
+		MQTT_TRC("Websocket connect failed (%d)",
+			 client->transport.websocket.sock);
+		return client->transport.websocket.sock;
+	}
+
+	MQTT_TRC("Connect completed");
+
+	return 0;
+}
+
+/**@brief Handles write requests on TCP socket transport.
+ *
+ * @param[in] client Identifies the client on which the procedure is requested.
+ * @param[in] data Data to be written on the transport.
+ * @param[in] datalen Length of data to be written on the transport.
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ */
+int mqtt_client_websocket_write(struct mqtt_client *client, const u8_t *data,
+				u32_t datalen)
+{
+	u32_t offset = 0U;
+	int ret;
+
+	while (offset < datalen) {
+		ret = websocket_send_msg(client->transport.websocket.sock,
+					 data + offset, datalen - offset,
+					 WEBSOCKET_OPCODE_DATA_BINARY,
+					 true, true, K_FOREVER);
+		if (ret < 0) {
+			return -errno;
+		}
+
+		offset += ret;
+	}
+
+	return 0;
+}
+
+/**@brief Handles read requests on TCP socket transport.
+ *
+ * @param[in] client Identifies the client on which the procedure is requested.
+ * @param[in] data Pointer where read data is to be fetched.
+ * @param[in] buflen Size of memory provided for the operation.
+ * @param[in] shall_block Information whether the call should block or not.
+ *
+ * @retval Number of bytes read or an error code indicating reason for failure.
+ *         0 if connection was closed.
+ */
+int mqtt_client_websocket_read(struct mqtt_client *client, u8_t *data,
+			       u32_t buflen, bool shall_block)
+{
+	int flags = 0;
+	int ret;
+
+	if (!shall_block) {
+		flags |= MSG_DONTWAIT;
+	}
+
+	ret = recv(client->transport.websocket.sock, data, buflen, flags);
```

Comment:
I think we should be consistent in terms of which API we use (`mqtt_client_websocket_write` calls `websocket_send_msg` while `mqtt_client_websocket_read` calls `recv`).

Personally, I'd stick to the `websocket_*` API, as from my point of view, it should make it easier to get this transport integrated with offloading. After moving the offloading underneath `zsock_*` (hopefully, I could look into this topic soon), both APIs should work, so I won't object if you decide to use socket API, but for now `websocket_*` is the only viable option for offloading.

Reply:
Agreed, I will change the code to use websocket_* API.

Note that the websocket API uses socket interface so the offloading case should be automatically handled.